### PR TITLE
add addcdiv.out, addcmul.out, baddbmm.out, cat.out etc; refactor op name according to native_functions.yaml

### PR DESCRIPTION
**Pull request by taozhiwei**

### CODE REVIEW

File: src/flag_gems/ops/addcmul.py [hunk 1]
```diff
@@ -16,15 +16,20 @@ def addcmul_forward(x, t1, t2, value):
     return x + value * t1 * t2
 
 
-def addcmul(inp, tensor1, tensor2, *, value=1.0, out=None):
+def addcmul_out(inp, tensor1, tensor2, *, value=1.0, out):
     logger.debug("GEMS ADDCMUL FORWARD")
```
Comment (tengqm) [inline src/flag_gems/ops/addcmul.py]:
```suggestion
    logger.debug("GEMS ADDCMUL_OUT")
```

File: src/flag_gems/ops/addcmul.py [hunk 2]
```diff
@@ -16,15 +16,20 @@ def addcmul_forward(x, t1, t2, value):
...
+    broadcast_shape = torch.broadcast_shapes(inp.shape, tensor1.shape, tensor2.shape)
+    if list(out.shape) != list(broadcast_shape):
+        out.resize_(broadcast_shape)
+    addcmul_forward(inp, tensor1, tensor2, value, out0=out)
+    return out
+
+
+def addcmul(inp, tensor1, tensor2, *, value=1.0):
+    """Functional entry; keep alongside ``addcmul.out`` for dispatch coverage."""
+    broadcast_shape = torch.broadcast_shapes(inp.shape, tensor1.shape, tensor2.shape)
```
Comment (tengqm) [inline src/flag_gems/ops/addcmul.py:31]: insert
```
logger.log("GEMS ADDCMUL")
```

Comment (tengqm) [reply src/flag_gems/ops/addcmul.py:31]: Please check other operators for similar nits.

File: src/flag_gems/ops/baddbmm.py [hunk 3]
```diff
@@ -234,6 +239,24 @@ def compute_B_grad(A, d_output, alpha):
     return grad_B
 
 
+def baddbmm_out(bias, A, B, *, beta=1.0, alpha=1.0, out):
+    logger.debug("GEMS BADDBMM OUT")
```
Comment (tengqm) [inline src/flag_gems/ops/baddbmm.py]:
```suggestion
    logger.debug("GEMS BADDBMM_OUT")
```

File: src/flag_gems/ops/log_softmax.py [hunk 4]
```diff
@@ -95,8 +95,8 @@ def log_softmax_backward_kernel(
         tl.store(in_grad_ptrs, in_grad, mask=mask)
 
 
-def log_softmax(self, dim, half_to_float=False):
-    logger.debug("GEMS LOG_SOFTMAX")
+def log_softmax_out(self, dim, half_to_float=False, *, out):
+    logger.debug("GEMS LOG_SOFTMAX OUT")
```
Comment (tengqm) [inline src/flag_gems/ops/log_softmax.py]:
```suggestion
    logger.debug("GEMS LOG_SOFTMAX_OUT")
```

File: src/flag_gems/ops/log_softmax.py [hunk 5]
```diff
@@ -128,8 +133,17 @@ def log_softmax(self, dim, half_to_float=False):
...
+    logger.debug("GEMS LOG_SOFTMAX")
+    assert dim >= -self.ndim and dim < self.ndim, "Invalid dim"
+    dim = dim % self.ndim
+    dtype = torch.float32 if half_to_float else self.dtype
+    out = torch.empty_like(self.contiguous(), dtype=dtype)
+    return log_softmax_out(self, dim, half_to_float, out=out)
+
+
+def log_softmax_backward_out(grad_output, output, dim, input_dtype, *, out):
+    logger.debug("GEMS LOG_SOFTMAX VJP OUT")
```
Comment (tengqm) [inline src/flag_gems/ops/log_softmax.py]: The operator targeted is `_log_softmax_backward_data.out`, so here we should log
`GEMS _LOG_SOFTMAX_BACKWARD_DATA_OUT`.

**Review by tengqm — COMMENTED** [review]
Please consider split this into several PRs, each with a single operator implemented along with test cases.

**Review by 0x45f — DISMISSED** [review]

File: tests/test_newly_registered_dispatch_ops.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
+"""Accuracy tests for aten overloads tied to recently registered FlagGems kernels."""
```
Comment (0x45f) [inline tests/test_newly_registered_dispatch_ops.py]: Move the unit tests to existing files

File: src/flag_gems/ops/attention.py [hunk 7]
```diff
@@ -1072,7 +1072,7 @@ def scaled_dot_product_attention(
     )
 
 
-def flash_attention_forward(
+def _flash_attention_forward(
```
Comment (tengqm) [inline src/flag_gems/ops/attention.py:1075]: Though I agree to most of the changes maded in this PR, which is a nice cleanup, I don't think we should change function names for this.
In Python's convention, a function name with a leading underscore is usually treated as a private function. We can and should make sure when we are calling `logger.debug` and when we are registering the operator, we use the operator names properly. Other than that, the function names do not matter that much.

We can finetune this protocol when we have both `_foo` and `foo` as two standalone operators. Before that, I don't changes like this in the PR is needed.

WDYT?

Comment (taozhiwei) [reply src/flag_gems/ops/attention.py:1075]: Personally, I think it would be better to follow the naming convention of native_functions.yaml. These registered ops are indeed not intended for users to directly invoke.
In addition, torch.cuda.memory._record_memory_history and torch.cuda.memory._dump_snapshot, etc.  also start with the  _.

**Comment by taozhiwei** [comment]
addmm.dtype_out and addmm.dtype is added in pytorch 2.8.0, should these two be deleted?

File: src/flag_gems/ops/attention.py (hunk 7 above)
Comment (tengqm) [reply src/flag_gems/ops/attention.py:1075]: That means when we register the operator in `flag_gems.__init__.py`, we would use 
`('_flash_attention_foward', flash_attention_forward_func),`. We can clearly state which target function we are replacing, but we don't have to use a leading underscore in the implementation.

A more general (maybe politically incorrect) idea is that when we are attempting to substitute `_foo` in torch, we are hacking. Maybe `_foo` was not designed to be a stable, exposed function. We are hacking because that is the real world we live in. However, we don't "hack" ourselves. Does this make senses to you?

**Comment by 0x45f** [comment]
> addmm.dtype_out and addmm.dtype is added in pytorch 2.8.0, should these two be deleted?addmm.dtype_out 和 addmm.dtype 是在 pytorch 2.8.0 中添加的，这两个函数应该删除吗？

keep these two operators

**Comment by taozhiwei** [comment]
> > addmm.dtype_out and addmm.dtype is added in pytorch 2.8.0, should these two be deleted?addmm.dtype_out 和 addmm.dtype 是在 pytorch 2.8.0 中添加的，这两个函数应该删除吗？
> 
> keep these two operators

I add skip in test

File: src/flag_gems/ops/addcdiv.py [hunk 8]
```diff
@@ -16,12 +16,14 @@ def addcdiv_kernel(x, t1, t2, value):
...
     addcdiv_kernel(inp, tensor1, tensor2, value, out0=out)
-
     return out
+
+
+def addcdiv(inp, tensor1, tensor2, value=1.0):
+    """Functional entry; CUDA may dispatch here without hitting ``addcdiv.out``."""
+    logger.debug("GEMS ADDCDIV")
+    out = torch.empty_like(inp)
+    return addcdiv_out(inp, tensor1, tensor2, value=value, out=out)
```
Comment (tengqm) [inline src/flag_gems/ops/addcdiv.py:29]: Why not direct calling addcdiv_kernel?

File: src/flag_gems/ops/log_softmax.py [hunk 9]
```diff
@@ -139,20 +153,31 @@ def log_softmax_backward(grad_output, output, dim, input_dtype):
...
             M,
             N,
             K,
         )
-    return in_grad
+    return out
+
+
+def log_softmax_backward(grad_output, output, dim, input_dtype):
+    logger.debug("GEMS LOG_SOFTMAX VJP")
```
Comment (tengqm) [inline src/flag_gems/ops/log_softmax.py:181]:
```suggestion
    logger.debug("GEMS LOG_SOFTMAX_BACKWARD")
```

File: src/flag_gems/ops/softmax.py [hunk 10]
```diff
@@ -281,31 +281,29 @@ def softmax_backward_kernel_inner(
             offsets += TILE_N
 
 
-def softmax(self, dim, half_to_float=False):
-    logger.debug("GEMS SOFTMAX")
+def softmax_out(self, dim, half_to_float=False, *, out):
+    logger.debug("GEMS SOFTMAX OUT")
```
Comment (tengqm) [inline src/flag_gems/ops/softmax.py:285]:
```suggestion
    logger.debug("GEMS SOFTMAX_OUT")
```

File: src/flag_gems/ops/softmax.py [hunk 11]
```diff
@@ -328,8 +326,24 @@ def softmax(self, dim, half_to_float=False):
...
+        zero_(out)
+        return out
+
+    dtype = torch.float32 if half_to_float else self.dtype
+    out = torch.empty_like(self, dtype=dtype)
+    return softmax_out(self, dim, half_to_float, out=out)
+
+
+def softmax_backward_out(grad_output, output, dim, input_dtype, *, grad_input):
+    logger.debug("GEMS SOFTMAX VJP OUT")
```
Comment (tengqm) [inline src/flag_gems/ops/softmax.py:346]:
```suggestion
    logger.debug("GEMS SOFTMAX_BACKWARD_OUT")
```

File: src/flag_gems/ops/softmax.py [hunk 12]
```diff
@@ -358,8 +377,16 @@ def softmax_backward(grad_output, output, dim, input_dtype):
...
+                grad_input,
                 M,
                 N,
             )
-    return in_grad
+    return grad_input
+
+
+def softmax_backward(grad_output, output, dim, input_dtype):
+    logger.debug("GEMS SOFTMAX VJP")
```
Comment (tengqm) [inline src/flag_gems/ops/softmax.py:388]:
```suggestion
    logger.debug("GEMS SOFTMAX_BACKWARD")
```

File: tests/test_blas_ops.py [hunk 13]
```diff
@@ -83,6 +84,66 @@ def test_addmm(M, N, K, scalar, dtype, b_column_major):
         del os.environ["MUSA_ENABLE_SQMMA"]
 
 
+@pytest.mark.addmm
```
Comment (tengqm) [inline tests/test_blas_ops.py:87]: Since we registered two operators as follows, this test must correspond to one of them. Either it is marked `@pytest.mark.addmm_dtype`, or `@pytest.mark.addmm_dtype_out`. Or else, we lose track of operator status.

```
("addmm.dtype", addmm_dtype),
("addmm.dtype_out", addmm_dtype_out),
```

File: tests/test_blas_ops.py [hunk 14]
```diff
@@ -83,6 +84,66 @@ def test_addmm(M, N, K, scalar, dtype, b_column_major):
...
+        res_out = torch.ops.aten.addmm.dtype(
+            bias, mat1, mat2, torch.float32, beta=1.0, alpha=1.0
+        )
+
+    gems_assert_close(
+        res_out, ref_out.to(flag_gems.device), torch.float32, reduce_dim=K
+    )
+
+
+@pytest.mark.addmm
```
Comment (tengqm) [inline tests/test_blas_ops.py:118]: Maybe you meant `@pytest.mark.addmm_dtype_out` for this test case.

File: tests/test_blas_ops.py [hunk 15]
```diff
@@ -350,6 +411,31 @@ def test_baddbmm_backward(M, N, K, scalar, dtype):
     gems_assert_close(res_in_grad2, ref_in_grad2, dtype, reduce_dim=M)
 
 
+@pytest.mark.baddbmm
```
Comment (tengqm) [inline tests/test_blas_ops.py:414]:
```suggestion
@pytest.mark.baddbmm_out
```

File: tests/test_special_ops.py [hunk 16]
```diff
@@ -1481,6 +1481,25 @@ def test_accuracy_cat_empty_tensor(shape, dim, dtype):
 )
 
 
+@pytest.mark.cat
```
Comment (tengqm) [inline tests/test_special_ops.py:1484]:
```suggestion
@pytest.mark.cat_out
```

File: tests/test_binary_pointwise_ops.py [hunk 17]
```diff
@@ -2533,6 +2533,28 @@ def test_accuracy_addcdiv(shape, dtype):
     gems_assert_close(res_out, ref_out, dtype)
 
 
+@pytest.mark.addcdiv
```
Comment (tengqm) [inline tests/test_binary_pointwise_ops.py:2536]:
```suggestion
@pytest.mark.addcdiv_out
```

File: src/flag_gems/__init__.py [hunk 18]
```diff
@@ -64,11 +68,15 @@ def torch_ge(v):
     ("add.Tensor", add),
     ("add_.Tensor", add_),
     ("addcdiv", addcdiv),
+    ("addcdiv.out", addcdiv_out),
     ("addcmul", addcmul),
+    ("addcmul.out", addcmul_out),
```
Comment (tengqm) [inline src/flag_gems/__init__.py:73]: where is the test case for this operator?

File: src/flag_gems/__init__.py [hunk 19]
```diff
@@ -42,10 +42,14 @@ def torch_ge(v):
...
     ("_is_all_true", _is_all_true),
     ("_log_softmax", log_softmax),
+    ("_log_softmax.out", log_softmax_out),
     ("_log_softmax_backward_data", log_softmax_backward),
+    ("_log_softmax_backward_data.out", log_softmax_backward_out),
     ("_safe_softmax", _safe_softmax),
     ("_softmax", softmax),
+    ("_softmax.out", softmax_out),
     ("_softmax_backward_data", softmax_backward),
+    ("_softmax_backward_data.out", softmax_backward_out),
```
Comment (tengqm) [inline src/flag_gems/__init__.py:52]: Where are the test cases for these new operators?

File: src/flag_gems/ops/__init__.py [hunk 20]
```diff
@@ -344,8 +354,12 @@
     "add",
     "add_",
     "addcdiv",
+    "addcdiv_out",
     "addcmul",
+    "addcmul_out",
```
Comment (tengqm) [inline src/flag_gems/ops/__init__.py:359]: This PR is a combination of changes to many operators.
It is hard to tell which operator is added, with or without accuracy/performance tests. For example, this `addcmul_out` operator doesn't have an accuracy or performance test ... this is not acceptable.

Please consider splitting this into smaller PRs, each focusing on a specific operator. It will help accelerate the PR review.

File: src/flag_gems/ops/attention.py (hunk 7 above)
Comment (taozhiwei) [reply src/flag_gems/ops/attention.py]: has changed

File: src/flag_gems/__init__.py (hunk 18 above)
Comment (taozhiwei) [reply src/flag_gems/__init__.py:74]: has added

File: src/flag_gems/ops/__init__.py (hunk 20 above)
Comment (taozhiwei) [reply src/flag_gems/ops/__init__.py:375]: has added

**Comment by tengqm** [comment]
@taozhiwei Please help check the merge conflicts.
You may also want to split this PR into smaller ones each focusing on a single operator (with test cases). With test case files split, you can now work in parallel on several operators without worrying that there will be merge conflicts on test case files.
For `foo.out` operator, please use the test case `test_foo.py` rather than creating a new test case file, though.

**Comment by tengqm** [comment]
@taozhiwei We are splitting the unit tests for easy of contributions and efficiency of CI pipeline. Sorry for the short-term inconvenience caused. Please help rebase this on to master. Thank you.

File: src/flag_gems/__init__.py (hunk 19 above)
Comment (taozhiwei) [reply src/flag_gems/__init__.py:53]: has added

**Comment by taozhiwei** [comment]
> @taozhiwei We are splitting the unit tests for easy of contributions and efficiency of CI pipeline. Sorry for the short-term inconvenience caused. Please help rebase this on to master. Thank you.

done

**Review by 0x45f — APPROVED** [review]

File: src/flag_gems/ops/__init__.py [hunk 21]
```diff
@@ -708,6 +726,8 @@
     "soft_margin_loss_out",
     "softmax",
     "softmax_backward",
+    "softmax_backward_out",
+    "softmax_out",
```
Comment (tengqm) [inline src/flag_gems/ops/__init__.py:730]: I believe most of the newly added operators do not have corresponding benchmark added ...
